Introduce callables.

dan-zheng opened the pull request:
Introduce callables: values that can be "called" like functions.
Add a new `call` declaration kind to represent callable methods.

["Introducing callables" pitch](https://forums.swift.org/t/pitch-introduce-static-callables/21732). Proposal coming soon.

dan-zheng committed:
Introduce callables.

Introduce callables: values that can be "called" like functions.
Add a new `call` declaration kind to represent callable methods.

dan-zheng commented on include/swift/AST/Decl.h:
```diff
@@ -5885,7 +5903,53 @@ AbstractStorageDecl::AccessorRecord::getAccessor(AccessorKind kind) const {
+///   }
+/// }
+/// \endcode
+class CallDecl final : public FuncDecl {
```
Conforming `CallDecl` to `FuncDecl` vs `AbstractFunctionDecl` is discussed in [this thread](https://forums.swift.org/t/question-what-should-calldecl-inherit-from/21815?u=dan-zheng).
It has a few nice properties - code sharing with `FuncDecl` is the primary one.

dan-zheng replied:
Committed to making `CallDecl` inherit from `FuncDecl`.

dan-zheng commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -3324,7 +3336,7 @@ parseIdentifierDeclName(Parser &P, Identifier &Result, SourceLoc &Loc,
                         tok ResyncT3, tok ResyncT4,
                         TokenProperty ResyncP1) {
-  if (P.Tok.is(tok::identifier)) {
+  if (P.Tok.is(tok::identifier) || P.Tok.is(tok::kw_call)) {
```
Added to support parsing `call` as an identifier in non-declaration contexts.
This enables `func call` and `call(...)` (apply expressions).

dan-zheng commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -5228,6 +5228,107 @@ Type ConstraintSystem::simplifyAppliedOverloads(
+
+    // If this type conforms to protocols, look up all of their `call`
+    // declaration requirements.
+    // FIXME: Fix this to handle overridden requirements, e.g. using
```
TODO/WIP: This logic for "getting all call declarations" doesn't work properly.
It should mimic/be unified with `lookupQualified`, as mentioned in [this post](https://forums.swift.org/t/question-what-should-calldecl-inherit-from/21815/9?u=dan-zheng) by @slavapestov.

dan-zheng replied:
Done in https://github.com/apple/swift/pull/23517/commits/94a384feb522c132c6ec2e64f50daa53c940683a.

To find all `call` declarations, `TC.lookupMember` is used, which internally calls `lookupQualified`. This approach can be back-ported to `@dynamicCallable` and `@dynamicMemberLookup` to reduce code dupe (and improve correctness, in the case of `@dynamicCallable`). I'm interested in doing that in a follow-up.

dan-zheng commented on lib/Demangling/Demangler.cpp:
```diff
@@ -2829,6 +2829,8 @@ NodePointer Demangler::demangleFunctionEntity() {
       Args = TypeAndMaybePrivateName; Kind = Node::Kind::Allocator; break;
     case 'c':
       Args = TypeAndMaybePrivateName; Kind = Node::Kind::Constructor; break;
+    case 'F':
```
TODO: Add mangling/demangling tests.
Currently, there are no SIL tests for `call` declarations.
I informally checked that demangling works via `-emit-silgen`.
The old remangler might not work - haven't look into how to invoke/test it yet.

dan-zheng commented on lib/Serialization/Serialization.cpp:
```diff
@@ -3409,6 +3410,69 @@ void Serializer::writeDecl(const Decl *D) {
+
+    auto contextID = addDeclContextRef(fn->getDeclContext());
+
+    unsigned abbrCode = DeclTypeAbbrCodes[FuncLayout::Code];
```
TODO: Create a `CallLayout` that is more specialized than `FuncLayout` (like `AccessorLayout`), dropping some fields like static spelling.

dan-zheng replied:
Deferred for now. `CallDecl` has only two less fields than `FuncDecl` (static location and static spelling.)

dan-zheng commented on test/Syntax/Parser/tree.swift.result:
```diff
@@ -1,8 +1,8 @@
-<s118><s163><s92><s13><NULL/><NULL/><t6>// REQUIRES: syntax_parser_lib
+<s119><s164><s93><s13><NULL/><NULL/><t6>// REQUIRES: syntax_parser_lib
```
I’m not sure why changing this test to pass was necessary. I wonder if changing this test is breaking.

dan-zheng commented on test/decl/call/protocol.swift:
```diff
@@ -0,0 +1,34 @@
+let s2 = S2()
+// TODO: Fix this by fixing `getCallDeclarations` logic in CSSimplify.cpp.
+// Need to handle overriding declarations; mimic/unify code with `lookupQualified`.
+// _ = s2()()()
```
This is super critical to fix: relevant TODO comment in CSSimplify.cpp.

dan-zheng replied:
Fixed in https://github.com/apple/swift/pull/23517/commits/94a384feb522c132c6ec2e64f50daa53c940683a.

johnfairh commented on include/swift/IDE/SyntaxModel.h:
```diff
@@ -97,6 +97,7 @@ enum class SyntaxStructureKind : uint8_t {
   EnumElement,
   TypeAlias,
   Subscript,
+  Call,
```
fyi will need a simple change to `ModelASTWalker::walkToDeclPre()` to actually emit this, and I think a possibly less-simple change to `SyntaxModelContext::SyntaxModelContext()` to classify `tok::kw_call` as identifer/keyword depending on context (this affects syntax highlighting).

dan-zheng committed:
Finish main callable functionality.

- Use uniform naming: either "`call` member" or "`call` declaration".
- Fix `call` member resolution during type-checking.
- Enable more `func` attributes on `call` declarations.
- Add tests.

dan-zheng commented:
The main callable functionality is done and tested: ready for review.
Code completion and editor features require some more work.

DougGregor commented:
@swift-ci build toolchain

tkremenek commented:
@swift-ci please build toolchain

dan-zheng committed:
Update clang version in update-checkout-config.json.

clang requires minor changes to support `call` declarations in Swift.

Note: the clang remote should be reverted back to "apple/swift-clang".

dan-zheng committed:
Minor edits.

- Parse "call" as identifier in patterns and enum case names. Add tests.
- Disallow `@warn_unqualified_access` on `call` declarations.
- Update failing attribute tests.

dan-zheng commented:
@DougGregor @tkremenek 

Sorry, building failed because this PR depends on some minor changes to clang to support `call` declarations: https://github.com/apple/swift-clang/pull/284

I temporarily updated the clang version in `utils/update_checkout/update-checkout-config.json` to branch `callable` of `dan-zheng/swift-clang` so that building will work. If there's a better way, please let me know.

I've verified that all swift tests pass locally. Retriggering toolchain build.

dan-zheng commented:
@swift-ci Please build toolchain

dan-zheng commented:
@swift-ci Please build toolchain

dan-zheng commented:
Investigating errors.

dan-zheng committed:
Merge branch 'master' into callable

dan-zheng commented:
@swift-ci Please build toolchain

dan-zheng commented:
@swift-ci Please build toolchain

dan-zheng commented:
@swift-ci Please build toolchain macOS

dan-zheng commented:
@swift-ci Please build toolchain macOS

dan-zheng commented:
@swift-ci Please benchmark

DougGregor commented:
@swift-ci please test source compatibility

DougGregor commented:
@swift-ci please test source compatibility

DougGregor commented:
FWIW, SwiftLint (from the source compatibility suite) is failing with this error:

```
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite-debug/project_cache/SwiftLint/Source/SwiftLintFramework/Rules/Style/MultilineFunctionChainsRule.swift:178:52: error: consecutive statements on a line must be separated by ';'
        return subcalls.flatMap { call -> [NSRange] in
                                                   ^
                                                   ;
```

dan-zheng commented:
Closing, as [SE-0253](https://github.com/apple/swift-evolution/blob/master/proposals/0253-callable.md) was [returned for revision](https://forums.swift.org/t/returned-for-revision-se-0253-static-callables/23290).
Redesign using `func call` as the name for call-syntax delegate methods: https://github.com/apple/swift/pull/24299
